File: cpp_solutions/945. Minimum Increment to Make Array Unique.cpp
class Solution {
public:
    int minIncrementForUnique(vector<int>& nums) {

        int ans = 0;
        set<int> global_seen;
        int i = 1;
        for(auto it : nums){
            global_seen.insert(it);
        }
        set<int> local_seen;
        sort(nums.begin(), nums.end());
        for(auto it : nums){
            if (local_seen.find(it) == local_seen.end()){
                local_seen.insert(it);
            } else {
                while (global_seen.find(i)!=global_seen.end() || i<=it){
                    i+=1;
                }
                ans+=(i-it);
                global_seen.insert(i);
            }
        }
        return ans;
        
    }
};

File: python3_solution_set2/945. Minimum Increment to Make Array Unique.py
class Solution:
    def minIncrementForUnique(self, nums: List[int]) -> int:
        ## [1,1,2,2,3,7]
        
        ## always move a repeating number to smallest non occuring number

        ## maintain a counter of seen elements and find the smallest not seen element
        
        ans = 0
        global_seen = set(nums)
        i = 1
        nums = sorted(nums)
        local_seen = set()

        for x in nums:
            if x not in local_seen:
                local_seen.add(x)

            else:
                while i in global_seen or i <= x:
                    i+=1
                #print(i,x)
                ans+=(i-x)
                global_seen.add(i)

        return ans
